### ISSUE TITLE
MultivaluedSection with custom pushrow

### ISSUE THREAD

**Issue by Claeysson** [xmartlabs/Eureka#1812]
I have a struct that store favorites name and URL:
```Swift
struct Favorite: Codable {
    var name: String
    var url: URL?
}
```
I store favorites in an array. I want to have a `MultivaluedSection` that lets the user add favorites. My idea is that when a user taps "add favorite" they will be taken to a `pushrow` where they can type in name and URL. When the user goes back i want the `MultivaluedSection` to contain the typed in Favorite. Is this possible, and if so, how?

**Comment by TheCyberMike** [comment]
re #1467 
I made a pretty simple change to Eureka in my app's fork to allow any Row to be used as the Add row.  It allows the default ButtonRow to be used as-is and handled automatically.  But it also allows any other row such as a PushRow to be used as the AddButtonProvider.

In my app, in multiple places, when an end-user presses the add button, I present them a popup list of choices they may add.  They choose one, and I add that choice to the MultiValuedSection.

The simple changes are in this forked Eureka commit in my app:
https://github.com/TheCyberMike/Eureka/commit/bfcba0dd04bf0d11cb6ba526235ae4c10c2d73fd

Using a PushRow as an AddButtonProvider is somewhat tricky, since the PushRow invokes a separate view controller to present the list, then returns to the original form's view controller.
  
1.  So you must be sure NOT to rebuild the original form and its MultiValuedSection when viewWillAppear() and viewDidAppear() are called upon returning from the PushRow view controller.

2. Also, the choice itself is made within the PushRow view controller.  The PushRow handles preserving that choice back from the PushRow view controller.  But the .onChange is invoked while the PushRow view controller is still active.  I used a DispatchQueue.main.async closure to handle deferring the form insert call to when the original form's view controller is active.

3. Be sure to use the [weak self]'s to prevent memory leaks.

Below is an adapted example of usage code from my eContact Collect app (https://github.com/TheCyberMike/eContactCollect-iOS), where things like languages and data entry fields and email accounts are chosen from pre-defined lists.

`self.mForm = form
form +++ MultivaluedSection(multivaluedOptions: [.Insert, .Delete, .Reorder],
       header: NSLocalizedString("Shown languages in order to-be-shown", comment:"")) { mvSection in
        mvSection.tag = "mvs_langs"
        mvSection.showInsertIconInAddButton = false
        mvSection.addButtonProvider = { [weak self] section in
            return PushRow<String>(){ row in
                // REMEMBER: CANNOT rebuild the form from viewWillAppear() ... the form must remain intact during the PushRow's view
                // controller life cycle for this to work
                row.title = NSLocalizedString("Select new language to add", comment:"")
                row.tag = "add_new_lang"
                row.selectorTitle = NSLocalizedString("Choose one", comment:"")
                row.options = langOptionsArray
                }.onChange { [weak self] chgRow in
                    // PushRow has returned a value selected from the PushRow's view controller;
                    // note our context is still within the PushRow's view controller, not the original FormViewController
                    if chgRow.value != nil {   // this must be present to prevent an infinite loop
                        guard let tableView = chgRow.cell.formViewController()?.tableView, let indexPath = chgRow.indexPath else { return }
                        DispatchQueue.main.async {
                            // must dispatch this so the PushRow's SelectorViewController is dismissed first and the UI is back at the main FormViewController
                            // this triggers multivaluedRowToInsertAt() below
                            chgRow.cell.formViewController()?.tableView(tableView, commit: .insert, forRowAt: indexPath)
                        }
                    }
            }
        }  // end of addButtonProvider
        mvSection.multivaluedRowToInsertAt = { [weak self] index in
            // a verified-new langRegion code was chosen by the end-user; get it from the PushRow
            let fromPushRow = self!.mForm!.rowBy(tag: "add_new_lang") as! PushRow<String>
            let langRegionCode:String = fromPushRow.value!
            
            // create a new ButtonRow based upon the new entry
            let newRow = self!.makeLangButtonRow(forLang: langRegionCode)
            fromPushRow.value = nil     // clear out the PushRow's value so this newly chosen item does not remain "selected"
            fromPushRow.reload()        // note this will re-trigger .onChange in the PushRow so must ignore that re-trigger else infinite loop
            return newRow               // self.rowsHaveBeenAdded() will get invoked after this point
        }
    }`

**Comment by mats-claassen** [comment]
#1843 Adds support to use any row type as add button. For your use case you will have to do some extra work as you want to push another view controller and then add the row with the result.
You should override `GenericMultivaluedSection` and change the `initialize` function to change the action of the add button. And your pushRow will be responsible for triggering the insertion of the new row

**Comment by mats-claassen** [comment]
Fixed in #1843
